### PR TITLE
[query/service] implement index_bgen

### PR DESCRIPTION
**Pull request by danking**

### CODE REVIEW

File: hail/python/hail/backend/service_backend.py [hunk 1]
```diff
@@ -444,8 +451,52 @@ async def inputs(infile, _):
...
+        async def inputs(infile, _):
+            await write_int(infile, ServiceBackend.INDEX_BGEN)
+            await write_str(infile, tmp_dir())
+            await write_str(infile, self.billing_project)
+            await write_str(infile, self.remote_tmpdir)
+            await write_int(infile, len(files))
+            for fname in files:
+                await write_str(infile, fname)
+            await write_int(infile, len(index_file_map))
+            for k, v in index_file_map.values():
```
Comment (daniel-goldstein) [inline hail/python/hail/backend/service_backend.py:483]:
```suggestion
            for k, v in index_file_map.items():
```

File: hail/python/hail/backend/service_backend.py [hunk 2]
```diff
@@ -444,8 +451,52 @@ async def inputs(infile, _):
...
+                await write_str(infile, k)
+                await write_str(infile, v)
+            if referenceGenomeName is None:
+                await write_bool(infile, False)
+            else:
+                await write_bool(infile, True)
+                await write_str(infile, referenceGenomeName)
+            await write_int(infile, len(contig_recoding))
+            for k, v in contig_recoding.values():
```
Comment (daniel-goldstein) [inline hail/python/hail/backend/service_backend.py:492]:
```suggestion
            for k, v in contig_recoding.items():
```

File: hail/src/main/scala/is/hail/backend/service/ServiceBackend.scala [hunk 3]
```diff
@@ -715,6 +744,63 @@ class ServiceBackendSocketAPI2(
...
+        while (i < nContigRecoding) {
+          val k = readString()
+          val v = readString()
+          contigRecoding(k) = v
+          i += 1
+        }
+        val skipInvalidLoci = readBool()
+
+        try {
+          val result = backend.indexBgen(
```
Comment (daniel-goldstein) [inline hail/src/main/scala/is/hail/backend/service/ServiceBackend.scala:784]: `result` in unused so might as well not assign

**Review by daniel-goldstein — CHANGES_REQUESTED** [review]
Small things

**Comment by tpoterba** [comment]
I think that for a lot of this kind of functionality -- index_bgen, vcf_metadata, etc -- it's going to be easier to just convert them to IR functions and then they hit the same endpoint.

**Comment by tpoterba** [comment]
for now, though, full steam with whatever's easiest.

File: hail/src/main/scala/is/hail/backend/service/ServiceBackend.scala (hunk 3 above)
Comment (danking) [reply hail/src/main/scala/is/hail/backend/service/ServiceBackend.scala:785]: I made indexBgen return the string instead for symmetry with the other methods.

File: hail/python/hail/backend/service_backend.py (hunk 1 above)
Comment (daniel-goldstein) [reply hail/python/hail/backend/service_backend.py:497]: I don't think this changed, did a commit go on the wrong branch?

Comment (daniel-goldstein) [reply hail/python/hail/backend/service_backend.py:497]: Unclear to me how this is passing tests, or else the types are wrong.

**Review by daniel-goldstein — CHANGES_REQUESTED** [review]

**Review by daniel-goldstein — APPROVED** [review]

**Pull request merged**